mflike/foreground.py

```diff
@@ -482,36 +632,282 @@ def _bandpass_construction(self, _initialize=False, **params):
+                            blT, blP = self.return_beams(exp, nu_ghz, 0.)
+
+                        if "tt" in self.requested_cls or "te" in self.requested_cls:
+                            trans_normT = trapezoid(
+                                bp[..., np.newaxis] * _cmb2bb(nub)[..., np.newaxis] * blT, nub, axis=0
+                            )
+                            ratioT = (
+                                bp[..., np.newaxis] * _cmb2bb(nub)[..., np.newaxis] * blT / trans_normT
+                            )
+                            self.bandint_freqs_T.append([nub, ratioT])
+
+                        if "te" in self.requested_cls or "ee" in self.requested_cls:
+                            trans_normP = trapezoid(
+                                bp[..., np.newaxis] *  _cmb2bb(nub)[..., np.newaxis] * blP, nub, axis=0
+                            )
+                            ratioP = (
+                                bp[..., np.newaxis] * _cmb2bb(nub)[..., np.newaxis] * blP / trans_normP
+                            )
+                            self.bandint_freqs_P.append([nub, ratioP])
 
         # fgspectra can't mix monofrequency with [nu, bp]. If one channel is mono-frequency then we
         # assume all of them and pass to fgspectra an array (not list!!) of frequencies
         if data_are_monofreq:
-            self.bandint_freqs = np.asarray(self.bandint_freqs)
+            self.bandint_freqs_T = np.asarray(self.bandint_freqs_T)
+            self.bandint_freqs_P = np.asarray(self.bandint_freqs_P)
             if self._initialized:
                 self.log.info("bandpass is delta function, no band integration performed")
         self._initialized = True
 
+    ###########################################################################
+    ## This part deals with beam functions, i.e. reading beam from file or
+    ## computing it as a Gaussian beam. We also have a function to compute
+    ## the correction expected for a Gaussian beam in case of bandpass shift
+    ## that should be applied to any beam profile
+    ###########################################################################
+    def _read_yaml_file(self, file_path):
+        import yaml
+
+        filename = "%s.yaml" % file_path
+        if not os.path.exists("%s.yaml" % file_path):
+            raise ValueError("File " + filename + " does not exist!")
+
+        with open(filename, "r") as f:
+            file = yaml.load(f, Loader=yaml.Loader)
+
+        return file
+
+    def _init_beam_from_file(self):
+        """
+        Reads the beam profile from an external file or the sacc file.
+        It has to be a dictionary ``{"{exp}_s0": {"nu": nu, "beams": array(freqs, ells+2)},
+        "{exp}_s2": {"nu": nu, "beams": array(freqs, ells+2)},...}``
+        including temperature and polarization beams.
+        """
+
+        if not self.beam_file:
+            # option to read beam from sacc.
+            try:
```

Comment:
I have changed it like that:
```
if not self.beam_file:
        # option to read beam from sacc
        if not bool(self.beams):
            raise ValueError("Beams not stored in sacc files!")
    else:
        self.beams = self._read_yaml_file(self.beam_file)
```
I also made sure all the checks regarding beams are done [after the initialization of mflike](https://github.com/simonsobs/LAT_MFLike/blob/922aebb8ee243c415e9a4e770935f848be9db26e/mflike/foreground.py#L519), which makes things so much simpler. And I added [these lines](https://github.com/simonsobs/LAT_MFLike/blob/922aebb8ee243c415e9a4e770935f848be9db26e/mflike/foreground.py#L771) to check that the beam has the correct shape (nfreqs, ells), comparing nfreqs with the len of the bandpass["nu"] array.
This is useful because there are cases in which the tracer.beam in the sacc file is not empty but contains a beam with shape (ells). If we are not reading beams from external files, i.e.
```
beam_profile:
      beam_from_file: null
```
, we would use the beam stored in the sacc, but this may have the wrong shape

Reply:
It's fine. Since `self.beams` is an dict (empty by default) coming from `mflike.py`, you do not need to cast it into `bool` just write `if not self.beams:`.

Reply:
Done